lib/ui/widgets/ata_map.dart

```diff
@@ -0,0 +1,181 @@
+/// 4. Calulate distance between current location and marked position
+/// Example code :
+///
+/// #admin user
+/// ```dart
+/// ATAMap(
+///    isMoveableMarker: true,
+///)
+/// ```
+/// #normal user
+/// ```dart
+/// ATAMap(
+///    titleMarkedPosition: 'Office Position',
+///    markedLat: 10.762622,
+///    markedLng:  106.660172,
+///)
+/// ```
+class ATAMap extends StatefulWidget {
+  final double centerMapLat;
+  final double centerMapLng;
+  final double markedLat;
+  final double markedLng;
+  final bool isMoveableMarker;
+  final double authRange;
+  final String titleMarkedPosition;
+
+  ATAMap(
+      {this.markedLat = 10.7440878,
+      this.markedLng = 106.7007886,
+      this.centerMapLat = 10.7440878,
+      this.centerMapLng = 106.7007886,
+      this.isMoveableMarker = false,
+      this.authRange = 100,
+      this.titleMarkedPosition = "Marked Position"});
+
+  @override
+  State<ATAMap> createState() => ATAMapState();
+}
+
+class ATAMapState extends State<ATAMap> {
+  double currentMarkedLat;
+  double currentMarkedLng;
+  double currentLocationLat;
+  double currentLocationLng;
+  bool isMapReady = false;
+  static const int TIMEOUT_PIN_MARKER_MAP_LOADED = 2;
+  static const double DEFAULT_ZOOM = 17;
+  CameraPosition get defaultCamera {
+    return CameraPosition(
+      target: LatLng(widget.centerMapLat, widget.centerMapLng),
+      zoom: DEFAULT_ZOOM,
+    );
+  }
+
+  Completer<GoogleMapController> _controller = Completer();
+  Set<Marker> _markers = Set();
+  Set<Circle> _circles = Set();
+  BitmapDescriptor markedLocationIcon;
+  @override
+  void initState() {
+    super.initState();
+    currentMarkedLat = widget.markedLat;
+    currentMarkedLng = widget.markedLng;
+    _setCustomMapIcons();
+  }
+
+  void _setCustomMapIcons() async {
+    markedLocationIcon = await BitmapDescriptor.fromAssetImage(ImageConfiguration(size: Size(128, 128)), 'assets/images/marked-location-icon.png');
+  }
+
+  Future<Position> getCurrentLocation() async => await Geolocator().getCurrentPosition();
+
+  void _goToCurrentLocation() async {
+    try {
+      Position currentPosition = await getCurrentLocation();
+      final GoogleMapController controller = await _controller.future;
+      controller.animateCamera(
+          CameraUpdate.newCameraPosition(CameraPosition(target: LatLng(currentPosition.latitude, currentPosition.longitude), zoom: DEFAULT_ZOOM)));
+      currentLocationLat = currentPosition.latitude;
+      currentLocationLng = currentPosition.longitude;
+    } on PlatformException catch (e) {
+      if (e.code == 'PERMISSION_DENIED') {
+        showDialog(
+            context: context,
+            builder: (BuildContext context) {
+              return AlertDialog(title: Text("PERMISSION_DENIED"), content: Text('GPS device is OFF, please turn ON !'));
+            });
+      }
+    }
+    setState(() {
+      if (_markers.length == 0) _addMarker(LatLng(currentMarkedLat, currentMarkedLng));

```

Comment:
_addMarker already have setState() inside, so no need to setState() before calling it